### PR TITLE
Clipper: Add ability to launch clipper window with shortcut

### PR DESCRIPTION
**Pull request by marcushill**
Related forum post:
- https://discourse.joplinapp.org/t/shortcut-hotkey-for-the-web-clipper/4877

Relevant docs:
- https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/commands

Not sure if you'll want to take this given the [discussion of a new webclipper](https://discourse.joplinapp.org/t/an-alternative-web-clipper-for-joplin/4904/), but the first bit was simple and already finished so I thought I'd open the PR. 

This change adds a command to the `manifest.json` for the web clipper which launches the webclipper rather than clicking on it. Because this is a WebExtensions feature and not something homegrown, users are able to change (or remove) the shortcut using native browser functionality.

It should theoretically be possible to take the default values and add commands for each individual piece of functionality ("Clip simplified page", "Clip Selection", etc), but I left that for another PR since I ended up not having as much time to work on this as I had hoped.

### CODE REVIEW

**Comment by tessus** [comment]
I like the shortcut, but how do people use the keyboard to actually clip something? The only thing it does is open the clipper, but then I have to use the mouse anyway.

**Comment by marcushill** [comment]
Yes, sorry. That's the problem I was mentioning. I meant to open this as a draft PR, but I apparently hit the wrong button.

**Comment by tessus** [comment]
No worries, all good. I just label it as WIP...

**Comment by tessus** [comment]
@marcushill thx for the update. Is it still WIP or can I remove the label?

**Comment by marcushill** [comment]
I think it should be okay to review. I've gone with the following decisions:
1. Two shortcuts are necessary to clip. One to open the window and a second one to actually clip.
2. I've left most of the shortcuts undefined because I couldn't come up with 6 reasonable ones and I think I read Chrome will only let you define 4 in the manifest anyway. The user can always go and define them themselves. (maybe we should add docs on how to do this?)
3. I left screenshots for later since it's a bit of a special case.

Thanks!

**Comment by tessus** [comment]
I haven't tested the code yet, so is it possible to select something else by cycling through the choices with `TAB` or the cursor keys and select the current one with `Enter` or `Space`?

**Comment by marcushill** [comment]
Assuming you mean select a different notebook, yes. I actually hadn't thought to test that until just now, but it turns out that the popup gets focus when it opens. At least on my machine, it's not super obvious which element currently has focus, but it is doable! It even works to add/edit tags.

**Comment by tessus** [comment]
I actually meant, if I could select all possible clickable elements with the keyboard.
Or is the use of the mouse still required for anything?

e.g. 1) select a different target notebook 2) move to `clip simplified page` 3) clip

**Comment by marcushill** [comment]
Unless you want to edit the default note title, the mouse is not required for anything you have a shortcut defined for. 

If we use the defaults as an example, to clip the complete page you have to do the following:
1. Press `Alt+Shift+J` to open the popup. This is to load the content scripts into the page and because most of the web-clipper logic is in the popup and not in the background script.
2. Press `Alt+Shift+C` to do the equivalent of clicking "Clip Complete" buttons.

If you want to edit the notebook in between, you can do that using keyboard navigation inside of the popup window since it has focus.

The title is not editable since the second shortcut skips the confirm step. The easiest solution to this problem would be to move it to before the "confirm" step. 

Also, it looks like the buttons in the webclipper can't be focused or at least their tab order is really weird. I could try to fix that or we could track it as a pre-existing bug.

**Comment by tessus** [comment]
Well, this is a problem. If I can only clip the complete page, this is rather problematic.

So here is the way it should work:

You use the key combination to bring up the popup. Then the focus should be on the first item. It should be marked the same way as it would be when hovering with the mouse over it (lighter blue). Then it should be possible to move to the next field by either using the `Tab` key or the `arrow down` key. (simplified -> complete -> complete (html) -> selection -> screenshot -> URL -> Notebook -> Tags -> simplified -> ....)
To select a focused item either `Space` or `Enter` should be used.

Basically it must be possible to use the clipper without a mouse.

**Comment by marcushill** [comment]
I pushed a fix for the buttons and made gifs to show what I was trying to describe.

## Working by tabbing through the popup:
Firefox:
ClipperWorkingFirefox-2020-01-11_06 31 25

Chrome:
ClipperWorkingChrome

## Working without tabbing through the popup
ClipperWorkingChangeShortcut

You'll have to take my word for it that the shortcut I pressed with `Alt+SHIFT+W`

No mouse needed in any of the three cases.

File: Clipper/joplin-webclipper/popup/src/bridge.js [hunk 1]
```diff
@@ -41,15 +42,21 @@ class Bridge {
...
 				this.dispatch({ type: 'CLIPPED_CONTENT_SET', content: content });
+
+				if (command.useDefaultSettings) {
+					const state = this.store_.getState();
+					content.parent_id = state.selectedFolderId;
+					if (content.parent_id) {
+						this.sendContentToJoplin(content);
+					}
+				}
 			}
```
Comment (laurent22) [inline Clipper/joplin-webclipper/popup/src/bridge.js]: I would prefer if each command has a well defined purpose. "clippedContent" is to set the clipped content within the extension, but shouldn't be used to also submit that content to Joplin directly. So is it possible to create a separate command, titled for example "sendContentToJoplin", which does the above? You might need to refactor the code a bit so as not to duplicate the content and response structures.

File: Clipper/joplin-webclipper/background.js [hunk 2]
```diff
@@ -76,3 +76,60 @@ browser_.runtime.onMessage.addListener(async (command) => {
...
+		browser_.tabs.query(options, (tabs) => {
+			resolve(tabs);
+		});
+	});
+}
+
+async function sendClipMessage(clipType) {
+	const tabs = await getActiveTabs();
+	if (!tabs || !tabs.length) {
+		console.warn('No active tabs');
```
Comment (laurent22) [inline Clipper/joplin-webclipper/background.js]: `console.error` or throw an exception please.

File: Clipper/joplin-webclipper/background.js [hunk 3]
```diff
@@ -76,3 +76,60 @@ browser_.runtime.onMessage.addListener(async (command) => {
...
+	}
+	if (message.name) {
+		browser_.tabs.sendMessage(tabId, message);
+	}
+}
+
+browser.commands.onCommand.addListener(function(command) {
+	// We could enumerate these twice, but since we're in here first,
+	// why not save ourselves the trouble with this convention
+	console.log('Rah');
```
Comment (laurent22) [inline Clipper/joplin-webclipper/background.js]: Please remove.

File: Clipper/joplin-webclipper/content_scripts/index.js [hunk 4]
```diff
@@ -266,12 +266,13 @@
 
 	async function prepareCommandResponse(command) {
 		console.info(`Got command: ${command.name}`);
+		const shouldSendToJoplin = command.shouldSendToJoplin || false;
```
Comment (laurent22) [inline Clipper/joplin-webclipper/content_scripts/index.js]: `!!command.shouldSendToJoplin` please

**Comment by laurent22** [comment]
Thanks @marcushill, it's looking good now. There's just one small changes which I'm going to make, and we can merge.

**Comment by tessus** [comment]
@laurent22 unfortunately the key combination never shows up anywhere.

Maybe we should add a sentence to the description / details (unter the web address in the details section). Something like:

Alt+Shift+J / Alt+Opt+J - shortcut to open the web clipper

image

**Comment by laurent22** [comment]
Yes sure, we can add this. By the way, for me the shortcuts don't seem to work actually.

**Comment by tessus** [comment]
I'm on macOS and `Opt+Shift+J` works perfectly. I made a mistake above with the second shortcut (I'll update it in a sec).
